clash-cores/src/Clash/Cores/UART.hs

```diff
@@ -0,0 +1,207 @@
+  -- ^ The word to transmit
+  -> ( Signal dom Bit -- The output bits for the tx line
+     , Signal dom Bool -- An acknowledgement that the word to transmit
+                       -- has been received
+     )
+uartTx mWord = let uartTx' = uartTxT @(Div (HzToPeriod baudRate) (DomainPeriod dom))
+               in unbundle $ register (1, False) $ mealy uartTx' UartTxIdle mWord
+
+-- | State used for the uart receiver
+data UartRxState (period :: Nat)
+   = UartRxIdle
+   -- ^ Waiting for a start bit
+   | UartRxStartBit
+   -- ^ Wait until we are at the center of the start bit
+       { uartRxHalfPeriodCounter :: Index (Div period 2) -- Counter that overflows
+                                                         -- at half a uart period
+       }
+   | UartRxDataBits
+   -- ^ Receive data bits
+       { uartRxWord :: BitVector 8 -- The data received
+       , uartRxPeriodCounter :: Index period -- Counter that overflows after a
+                                             -- full uart period
+       , uartRxWordCounter :: Index 9 -- Which bit of uart we are currently
+                                      -- waiting for. This is 9 because we also
+                                      -- wait until we check the stop bit.
+       }
+   | UartRxStopBit
+   -- ^ Check the stop bit is correct
+       { uartRxWord :: BitVector 8 -- The completely received word
+       }
+  deriving (Show, Generic, NFDataX)
+
+-- State transition function for the uart receiver
+uartRxT
+  :: forall (period :: Nat)
+   . KnownNat period
+  => UartRxState period
+  -- ^ The current state
+  -> Bit
+  -- ^ The uart receive line
+  -> ( UartRxState period -- The next state
+     , Maybe (BitVector 8) -- A complete word that has been received
+     )
+uartRxT UartRxIdle 1 = (UartRxIdle, Nothing)
+uartRxT UartRxIdle 0 = (UartRxStartBit 1, Nothing)
+uartRxT UartRxStartBit{..} b = (nextState, Nothing)
+  where
+    halfPeriodDone = uartRxHalfPeriodCounter == maxBound
+    stillLow = b == 0
+    validStartBit = halfPeriodDone && stillLow
+
+    nextState | validStartBit  = UartRxDataBits
+                                   { uartRxWord = 0
+                                   , uartRxPeriodCounter = 0
+                                   , uartRxWordCounter = 0
+                                   }
+              | halfPeriodDone = UartRxIdle
+              | otherwise      = UartRxStartBit
+                                   { uartRxHalfPeriodCounter = uartRxHalfPeriodCounter + 1 }
+uartRxT state@UartRxDataBits{..} b = (nextState, Nothing)
+  where
+    periodDone = uartRxPeriodCounter == maxBound
+    lastBit = uartRxWordCounter == maxBound
+    finishedReceive = periodDone && lastBit
+
+    nextState | finishedReceive = UartRxStopBit { uartRxWord = uartRxWord }
+              | periodDone      = UartRxDataBits
+                                     { uartRxWord = replaceBit (7 :: Index 8)
+                                                      b (shiftR uartRxWord 1)
```

Comment:
We should have a [`+>>`](https://hackage.haskell.org/package/clash-prelude-1.4.2/docs/Clash-Explicit-Prelude.html#v:-43--62--62-) for bitvector.

Reply:
I changed the internal representation to a Vec and convert at the interface boundary.

Reply:
Ah sorry for the confusion, I wasn't suggesting you'd change anything. The version on bitvectors is probably better in terms of simulation performance (both memory and cpu) - so for now that's preferable. I was just saying (worded poorly) that I think we should add `+>>`/`<<+`/..  equivalents for bitvectors at some point.

Reply:
I have reverted it. In a perfect world `Vec n Bit` and `BitVector n` would really be the same.